db/db_statistics_test.cc

```diff
@@ -283,6 +284,71 @@ TEST_F(DBStatisticsTest, BlockChecksumStats) {
             options.statistics->getTickerCount(BLOCK_CHECKSUM_MISMATCH_COUNT));
 }
 
+TEST_F(DBStatisticsTest, BytesWrittenStats) {
+  Options options = CurrentOptions();
+  options.statistics = ROCKSDB_NAMESPACE::CreateDBStatistics();
+  options.statistics->set_stats_level(StatsLevel::kExceptHistogramOrTimers);
+
+  // Destroy the DB to recreate as a TransactionDB.
+  Destroy(options, true);
+
+  // Create a TransactionDB.
+  TransactionDB* txn_db = nullptr;
+  TransactionDBOptions txn_db_opts;
+  txn_db_opts.write_policy = TxnDBWritePolicy::WRITE_COMMITTED;
+  ASSERT_OK(TransactionDB::Open(options, txn_db_opts, dbname_, &txn_db));
+  ASSERT_NE(txn_db, nullptr);
+  db_ = txn_db->GetBaseDB();
+
+  EXPECT_EQ(0, options.statistics->getAndResetTickerCount(WAL_FILE_BYTES));
+  EXPECT_EQ(0, options.statistics->getAndResetTickerCount(BYTES_WRITTEN));
+
+  const int kNumKeysWritten = 100;
+  const int kCommitMarkerSize = 12;
```

Comment:
Can you elaborate a bit more how is this marker size calculated?

Reply:
Hi, yeah for sure. So I'm not too familiar with the codebase yet, but I think this comes from `WriteBatch::MarkWalTerminationPoint()`, which gets called in `WriteCommittedTxn::CommitInternal()`.
https://github.com/facebook/rocksdb/blob/2045fe4693dd8957796ccce3a4abbf3f689226b3/db/write_batch.cc#L285-L289

This `wal_term_point_` is of type `SavePoint`, which has these three fields. 
https://github.com/facebook/rocksdb/blob/2045fe4693dd8957796ccce3a4abbf3f689226b3/include/rocksdb/write_batch.h#L45-L48

The two `uint32_t` fields should be 4 bytes each, while `size_t` I think should be 4 bytes on a 32 bit machine, 8 bytes on a 64 bit machine. So in total that's either 12 or 16 bytes.

And through experimentation, I tested with different amount of entries written in the transaction, at the end the diff between `WAL_FILE_BYTES` and `BYTES_WRITTEN` was always 12, which is somewhat consistent with the size of the `SavePoint` struct (I tested on an x86_64 machine, but that extra size is 12 instead of 16, maybe some compiler optimization?).

Reply:
OK, thanks for the explanation. I think `SavePoint` is just an in memory data structure, `WriteBatchInternal::Count(WriteBatch)` would only count the encoded representation, so this in memory data structure is not included.

I checked a bit more on this, and figured that the extra 12 bytes come from an extra header written to WAL compared to the bytes passed to memtable. So it's the header size https://github.com/facebook/rocksdb/blob/2045fe4693dd8957796ccce3a4abbf3f689226b3/db/write_batch_internal.h#L81
I would recommend you to directly use this `WriteBatchInternal::kHeader` variable and add some documentation to explain it a little bit.

 I think this is where the difference comes from:
Step 1) `Transaction::Prepare`
writes to WAL: header + original_batch

Step 2) `Transaction::Commit`
writes to WAL:  header + commit_marker
while writes to Memtable:  header + commit_marker  + original_batch

(It achieves this with the above `MarkWalTerminationPoint` by skipping the `original_batch` when writing to WAL).

So in total WAL write has an extra header.

Reply:
Ah i see, thanks for clarifying! Yeah i'll just use that constant in the test.